### PR TITLE
Some code cleaning

### PR DESCRIPTION
**Pull request by yeesian**
* switch all occurrences of `Array` (which is an abstract type) to `Vector` or `Matrix` for clarity
* make `distcost` a keyword argument
* remove all the complicated indexing stuff for the OD matrix

### CODE REVIEW

**Comment by yeesian** [comment]
We need to list the packages that we need in REQUIRE files, so that they will be installed when `Pkg.add("TransitNetworks")` is run on other machines.

You can have an additional `REQUIRE` file in the `test/` folder if there are packages that you need only for the tests, but not for the package itself. Then if they aren't already installed, they will be installed at the start of the test and removed at the end. ([travis log](https://travis-ci.org/juliayyan/TransitNetworks.jl/jobs/360359508#L468-L476); [corresponding commit](https://github.com/juliayyan/TransitNetworks.jl/pull/2/commits/93bc6da69215961f16cb92c6fc228879de6a3dc8))

**Comment by yeesian** [comment]
For what it's worth, I think only the NetworkProblem stuff can be automated in this way, and we'll need a separate folder/package for the models, since Travis won't have a copy of Gurobi etc.

It makes sense to me, since we generally don't expect the code for the network representation to change frequently, and we expect it to be robust in handling all sorts of edge cases so that it's safe for us to use in our models.

On the other hand, we'll be iterating on the models a lot more, and having a lot of other packages that we'll like to rely on. So the testing of that component of the project will probably be kept local for academic research like we currently do.

**Comment by juliayyan** [comment]
sounds good

**Comment by yeesian** [comment]
From the logs: https://travis-ci.org/juliayyan/TransitNetworks.jl/builds/360371583

- it's currently working for OS X with julia v0.6.
- it's failing on Linux machines (log [here](https://travis-ci.org/juliayyan/TransitNetworks.jl/jobs/360371584#L591-L619)) because HDF5 (which is a dependency of `JLD`) isn't being properly installed. It's an upstream issue, and nothing to do with this package.

If we want it to be tested on windows, we'll need to do a similar setup for https://www.appveyor.com/.

**Comment by yeesian** [comment]
The `.travis.yml` file contains the configuration/instructions for what should be done on Travis.

In [this commit](https://github.com/juliayyan/TransitNetworks.jl/pull/2/commits/66ef3dd4cd3d4a83581a7ee6af3d6d5daac7cc43), I'm telling github to ignore the failures on
- julia nightlies (which are the [nightly builds](https://julialang.org/downloads/nightlies.html) of julia binaries; they're usually "unstable" since they contain fresh commits and are not certified for widespread distribution yet), and
- Linux (since I have very little motivation to fix the upstream problems with HDF5 for now).

**Comment by yeesian** [comment]
It'll help us to have a code coverage tool. To do that, set up an account on https://coveralls.io/ and set it up for the package.

**Comment by juliayyan** [comment]
Okay, I added the repo.

**Comment by yeesian** [comment]
Thanks! I have a church service now but I'll get back to it after

**Comment by yeesian** [comment]
Okay this PR is ready for review. I'm leaving out most of the tests, and will slowly introduce them in subsequent PRs after looking at the coverage reports.

**Comment by yeesian** [comment]
I think I'll introduce some data files within this package for testing purposes. I was hesitant at first because of pre-emptive fears of introducing bloat (see discussion [here](https://github.com/yeesian/ArchGDAL.jl/issues/10)).

But the way it is currently set up is not ergonomic for us, because our only way of interacting with the tests is through `Pkg.test("TransitNetworks")`.

File: src/network/networkproblem.jl [hunk 1]
```diff
@@ -530,9 +530,9 @@ function NetworkDesignProblem{R <: Real}(
     # (i) latlngs
     # -----
     coords = zeros(nstns, 2)
-    if nrow(latlngs) > 0
+    if DataFrames.nrow(latlngs) > 0
         latlngs = latlngs[findin(latlngs[:stop_id], stns),:]
-        for i in 1:nrow(latlngs)
+        for i in 1:DataFrames.nrow(latlngs)
```
Comment (juliayyan) [inline src/network/networkproblem.jl]: Just realized, I think we should have a test around here to check that all of the latlngs are present in the DataFrame (verify that all stns are represented in the stop_ids).

File: test/multiline-construction.jl [hunk 2]
```diff
@@ -4,19 +4,26 @@ module MultiLineConstruction
     println("Testing MultiLineConstruction")
     println("=============================")
 
-    using Base.Test, DataFrames, JLD
-    using TransitNetworks, LightGraphs
+    using Base.Test, DataFrames, TransitNetworks
+
+    demanddata = DataFrames.readtable(
+        "testdata/scenarios/year2014month2day4hour17to19.csv"
```
Comment (juliayyan) [inline test/multiline-construction.jl]: This is *great*

File: src/network/networkproblem.jl (hunk 1 above)
Comment (yeesian) [reply src/network/networkproblem.jl]: My understanding's that the filtering is done in L534:

```julia
latlngs = latlngs[findin(latlngs[:stop_id], stns),:]
```

such that by the time we get to this loop, the latlngs have already been filtered to those stops that are in `stns`.

**Comment by juliayyan** [comment]
I meant the other direction — we want to check that all of the stns are
present in the latlngs table.

**Comment by yeesian** [comment]
> I meant the other direction — we want to check that all of the stns are present in the latlngs table.

Should it be a warning or an error if not all of them are present in `latlngs`?

**Comment by juliayyan** [comment]
I think an error? I think latlngs should be required information (which is
why I moved them out of the keyword arguments).

**Comment by yeesian** [comment]
We get

```julia
==================================
Testing MultiRoutes (Main Network)
==================================
ERROR: LoadError: LoadError: String["1", "10003", "101", "104", "10590", "106", "107", "108", "109", "2", "57", "58", "6", "64", "74624", "87", "88", "89", "91", "95", "96", "97", "99"] not found in latlngs table
```

**Comment by juliayyan** [comment]
Yeah, it's because of this line

https://github.com/juliayyan/TransitNetworks.jl/blob/cleanup/test/multiroute-construction.jl#L76

which we should take out.

**Comment by yeesian** [comment]
What does the latlngs affect? based on https://github.com/juliayyan/TransitNetworks.jl/blob/244b6e34943a5af4ad29f9153407f95c4678bd12/src/network/networkmethods.jl#L42-L49
is it just for the taxis?

Seems a little stringent to me to say that the user shouldn't be able to build the transitnetwork just because they don't have all of the latlngs.

**Comment by yeesian** [comment]
> I was thinking that the costs for network design would be based on the distances between stations (calculated using the haversine distance)

That check can be enforced in the code when building models where it matters. But I'm thinking we shouldn't building any models in this package.

This same `TransitModels` package should eventually be a dependency and work for both `SchedulingModels` and (a yet-to-be-created) `NetworkDesignModels` package. What do you think?

**Comment by juliayyan** [comment]
Ah okay, didn't realize that we weren't building models here.

How will we deal with the fact that odmatrix is nstns x nstns x nperiods in SchedulingModels, and nstns x nstns here (for network design)?

**Comment by yeesian** [comment]
> How will we deal with the fact that odmatrix is nstns x nstns x nperiods in SchedulingModels, and nstns x nstns here (for network design)?

https://github.com/juliayyan/TransitNetworks.jl/pull/2/commits/09a9ebbe2c9723674bb9269f8145a40546418f37

**Comment by juliayyan** [comment]
Okay, nothing more on my end then.

**Comment by yeesian** [comment]
Actually, what do you think if we rename `NetworkDesignProblem` to `TransitNetwork`?

**Comment by juliayyan** [comment]
TransitNetworkProblem sounds good to me

**Pull request merged**